libpam/pam_handlers.c

```diff
@@ -576,37 +588,28 @@ int _pam_init_handlers(pam_handle_t *pamh)
  * preceded by lines of comments and also extended with "\\\n"
  */
 
-static int _pam_assemble_line(FILE *f, char *buffer, int buf_len)
+static int _pam_assemble_line(FILE *f, char **buffer, size_t buf_len)
 {
-    char *p = buffer;
-    char *endp = buffer + buf_len;
+    char *p;
     char *s, *os;
     int used = 0;
 
     /* loop broken with a 'break' when a non-'\\n' ended line is read */
 
     D(("called."));
     for (;;) {
-	if (p >= endp - 1) {
-	    /* Overflow */
-	    D(("overflow"));
-	    return -1;
-	}
-	if (fgets(p, endp - p, f) == NULL) {
+	if (getline(buffer, &buf_len, f) <= 0) {
 	    if (used) {
 		/* Incomplete read */
+		free(*buffer);
```

Comment:
This leads to a double free when the while-loop in _pam_parse_conf_file is left. The free is performed after the while as well.